lib/src/config_provider/config.dart

```diff
@@ -215,363 +219,689 @@ class Config {
     }
   }
 
-  /// Validates Yaml according to given specs.
-  bool _checkConfigs(YamlMap map, Map<List<String>, Specification> specs) {
-    var result = true;
-    for (final key in specs.keys) {
-      final spec = specs[key];
-      if (checkKeyInYaml(key, map)) {
-        result = result && spec!.validator(key, getKeyValueFromYaml(key, map));
-      } else if (spec!.requirement == Requirement.yes) {
-        _logger.severe("Key '$key' is required.");
-        result = false;
-      } else if (spec.requirement == Requirement.prefer) {
-        _logger.warning("Prefer adding Key '$key' to your config.");
-      }
+  Schema _getRootSchema() {
+    return FixedMapSchema(
+      keys: [
+        FixedMapKey(
+          key: strings.llvmPath,
+          valueSchema: ListSchema<String>(
+            childSchema: StringSchema(),
+            transform: (node) => llvmPathExtractor(node.value),
+          ),
+          defaultValue: (node) => findDylibAtDefaultLocations(),
```

Comment:
This is not part of the schema, but is part of the config.

So in some sense we could have some kind of `ConfigSpecification` abstraction that extends/implements `Schema` or has a method that returns a `Schema`. If we go for a method that returns the `Schema`, we need two class hierarchies that mirror each other, that would be verbose. If we go for `extends` we end up with the issue that all the child nodes only have a `Schema` type and not a `ConfigSpecification` type. If we go for `implement`s then all the `Schema` abstract classes would have abstract getters for their children (but that still means two type hierarchies).

Maybe the best thing is to call the whole thing `ConfigSpecification` instead of `Schema`? A config specification needs to specify the data structure (schema), but can specify more. And then our config schema currently just happens to mirror the different schema options and naming conventions.

The naming would then be something like:

* `Schema` -> `ConfigSpecification`
* `FixedMapSchema` -> `MapConfigSpecification` (with an optional argument `Iterable<K>? validKeys`, if no list is provided, you can have arbitrary keys, if a list is provided it's a fixed map?)
* `ListSchema<V>` -> `ListConfigSpecification<V>` and we can restrict values with an optional `Iterable<V>? validValues` argument to make it consistent with `Config.string` from `package:cli_config`.
* `FixedMapKey` -> `MapConfigSpecificationEntry<K,V>`
* `OneOfSchema` -> `OneOfConfigSpecification`

Reply:
How about `ConfigSpec` instead? a tad little less to type.

Reply:
I've used `ConfigSpec` for now. 

> MapConfigSpecification (with an optional argument Iterable<K>? validKeys, if no list is provided, you can have arbitrary keys, if a list is provided it's a fixed map?)

This works, but we lose the ability to validate different key patterns with different types of schema (albeit we don't yet have a usecase of different patterns with different schemas, but we might in the future)

Reply:
<img width="645" alt="image" src="https://github.com/dart-lang/ffigen/assets/3601775/cda27f89-c83b-47ae-971d-c6b37392ff0c">

`ConfigDefinition` would be shorter. Or `ConfigStructure`. Or, .. we call this thing the `Config` and we call the instantiated one the `ConfigInstance`.

@lrhn to the rescue? 😄

Reply:
>ListConfigSpecification<V> and we can restrict values with an optional Iterable<V>? validValues argument to make it consistent with Config.string from package:cli_config

We can use `ListSchema(childSchema: EnumSchema(allowedValues: [...]))`